### PR TITLE
Simpler AddressParsing by making AddressParser a @FunctionalInterface

### PR DESCRIPTION
**Pull request by msgilligan**
The main idea here is that we made `AddressParser` the `@FunctionalInterface` and a nested class is used to define an `AddressParserProvider` (Instead of the other way around.)

### CODE REVIEW

**Comment by schildbach** [comment]
Yes, it seems to me this is the better approach.

**Comment by msgilligan** [comment]
@schildbach I renamed `DefaultAddressParser` to `DefaultAddressParserProvider` and made it package-private. I also add some missing JavaDoc and force-pushed. It's now ready for final review and merging.

**Comment by schildbach** [comment]
Merged.

File: core/src/main/java/org/bitcoinj/base/AddressParser.java [hunk 1]
```diff
@@ -17,42 +17,93 @@
...
+     * @return A parser that will throw for strings that are not valid for network.
+     */
+    @Deprecated
+    static AddressParser getLegacy(@Nullable NetworkParameters params) {
+        AddressParser.AddressParserProvider provider = DefaultAddressParserProvider.fromNetworks();
+        return (params == null)
+                ? provider.forKnownNetworks()
+                : provider.forNetwork(params.network());
     }
+
```
Comment (Rjsoon38) [inline core/src/main/java/org/bitcoinj/base/AddressParser.java:108]: Would you place a different type of command here if you needed to? Such as if you needed to make changes or additions to a different type of command?

**Pull request closed without merging**